src/Common/FileSegment.cpp

```diff
@@ -471,144 +472,78 @@ void FileSegment::completeBatchAndResetDownloader()
 
 void FileSegment::complete(State state)
 {
-    std::lock_guard cache_lock(cache->mutex);
     std::lock_guard segment_lock(mutex);
 
-    assertNotDetached(segment_lock);
-
-    bool is_downloader = isDownloaderImpl(segment_lock);
-    if (!is_downloader)
     {
-        cv.notify_all();
-        throw Exception(ErrorCodes::REMOTE_FS_OBJECT_CACHE_ERROR,
-                        "File segment can be completed only by downloader or downloader's FileSegmentsHodler");
-    }
+        SCOPE_EXIT({
+            download_state = state;
```

Comment:
Why do you need to set it on SCOPE_EXIT and not instantly (`download_state = state`)